### ISSUE TITLE
regression: login breaks for accounts with uppercase username

### ISSUE THREAD

**Issue by pedro-nonfree** [matrix-org/matrix-synapse-ldap3#167]
### Description

co-written with @evilham

Our ldap matrix instance did ldap case insensitive login which means that it was unaffected by https://github.com/matrix-org/matrix-synapse-ldap3/issues/165

### Steps to reproduce

With synapse [1.62.0](https://github.com/matrix-org/synapse/releases/tag/v1.62.0) [this change](https://github.com/matrix-org/synapse/issues/13156) in matrix-synapse-ldap3 breaks ldap login for similar setups when historical users have uppercase letters.

This is a very old installation before https://spec.matrix.org/v1.3/appendices/#user-identifiers and it does not seem to be a migration to lowercase usernames.

### Homeserver

selfhosted homeserver instance

### Synapse Version

{"server_version":"1.62.0","python_version":"3.9.2"}

### Installation Method

Debian packages from packages.matrix.org

### Platform

Debian 11 Linux

### Relevant log output

If you want to look at `homeserver.log` there are in https://github.com/vector-im/element-web/issues/22859

### Anything else that would be useful to know?

This is what we get on our database on a login attempt:

```
synapse=# select name from users where name='@SomeUser:matrix.example.com';
           name
---------------------------
 @SomeUser:matrix.example.com
(1 row)

synapse=# select user_id from access_tokens where id=TOKEN_ID;
          user_id
---------------------------
 @someuser:matrix.example.com
(1 row)
```

This inconsistency means that the `INNER JOIN` in this query fails and the login proceeds with trying the macaroon which is not going to work

On this line:

https://github.com/matrix-org/synapse/blob/96cf81e312407f0caba1b45ba9899906b1dcc098/synapse/storage/databases/main/registration.py#L547

And trying to fix the query with `LOWER(users.name)` is not enough. Client fails on `filters` endpoint with error `403` with response `{"errcode":"M_FORBIDDEN","error":"Cannot create filters for other users"}`. Other endpoints might have similar issues

As a temporary workaround we are rolling back the python package matrix-synapse-ldap3 this way (it's a debian package):

```
service matrix-synapse stop
/opt/venvs/matrix-synapse/bin/pip install matrix-synapse-ldap3==0.2.0
service matrix-synapse start
```

As shown in reproducing the bug, looks like it is a problem related to how synapse server deals with case insensitive user IDs and not to the matrix-synapse-ldap3 python package.

**Comment by H-Shay** [comment]
Hello and thanks for reporting this. Just so I am clear (it's a little hard to follow the info across the issues), you upgraded to synapse 1.62.0, and after that login attempts for users with old, historical usernames containing capital letters began to fail, correct? 
This is part I am a little unclear on: can you let me know exactly what the failure looks like? I.e. what error message you receive from the client and the relevant logs from the homeserver covering the login request, that would be extremely helpful. If you feel uncomfortable pasting the logs here please feel free to DM them to me at @shayshay:matrix.org.

**Comment by H-Shay** [comment]
Thanks for the info-I just still have one question, which is: what error message/error code are you seeing when the request to `/login` fails?

**Comment by evilham** [comment]
This meaningless thing:

```
2022-07-15 14:19:40,194 - synapse.storage.database - 803 - WARNING - sentinel - Starting db txn 'add_access_token_to_user' from sentinel context
2022-07-15 14:19:40,280 - synapse.api.auth - 450 - WARNING - GET-472531 - Invalid access token in auth: <class 'pymacaroons.exceptions.MacaroonDeserializationException'> cannot determine data format of binary-encoded macaroon.
2022-07-15 14:19:40,280 - synapse.http.server - 167 - INFO - GET-472531 - <SynapseRequest at 0x7f024d753d30 method='GET' uri='/_matrix/client/unstable/org.matrix.msc2697.v2/dehydrated_device?&_cacheBuster=6499453651040419' clientproto='HTTP/1.0' site='8008'> SynapseError: 401 - Invalid access token passed.
2022-07-15 14:19:40,455 - synapse.api.auth - 450 - WARNING - POST-472533 - Invalid access token in auth: <class 'pymacaroons.exceptions.MacaroonDeserializationException'> cannot determine data format of binary-encoded macaroon.
2022-07-15 14:19:40,455 - synapse.http.server - 167 - INFO - POST-472533 - <SynapseRequest at 0x7f027074aee0 method='POST' uri='/_matrix/client/r0/keys/upload?' clientproto='HTTP/1.0' site='8008'> SynapseError: 401 - Invalid access token passed.
```

As described, it tries the regular flow, then goes on with the macaroons, which ofc can't be deseralised because it's been ages since that got retired, so the result is: the token is not valid, aka https://http.cat/401

**Comment by H-Shay** [comment]
Great thanks!

**Comment by H-Shay** [comment]
We have landed a fix in https://github.com/matrix-org/matrix-synapse-ldap3/pull/168 which should address the issue.
